### ISSUE TITLE
Add "sync now" button

### ISSUE THREAD

**Issue by aidan-fitz** [ascoderu/opwen-webapp#68]
It would be useful to connect the Lokole to the cell network during emergencies, such as natural disasters, regardless of time of day. This could be enabled manually by the device admin or remotely via some signal, such as SMS or LoRa.

**Comment by nzola** [comment]
This is a great idea. The Lokole is an offline device and it is normally not connected on Internet. You are proposing to implement the feature which can turn the Lokole on real time live connected on cellphone? The feature to be used in emergency. Is this correct?

**Comment by aidan-fitz** [comment]
Yes, that's what I'm thinking.

**Comment by nzola** [comment]
Last week, someone was talking about this to me. His name is Mic. I will invite him to this conversation.

**Comment by c-w** [comment]
This is a good idea, @aidan-fitz. We've discussed a similar feature for the admin UI ("sync now" button) in the past.

The main reason that the real-time functionality is not currently implemented is that it'd require a few changes in the Linux setup, mostly to enable the user that runs the webapp to have all the required permissions to dial up via usb_modeswitch and wvdial (currently this is done via a root cron job which works around the permission issues in a rather crude way).

The current flow is this:

![Diagram showing existing sync flow](https://user-images.githubusercontent.com/1086421/47434658-fa45fd80-d770-11e8-9a3d-06c7a03b4efb.png)

To enable a more "on demand" online mode, we'd need to change the flow to this:

![Diagram showing on-demand sync flow](https://user-images.githubusercontent.com/1086421/47434720-25305180-d771-11e8-9228-a6ced2427fe6.png)

The main change here is that the opwen_webapp now calls usb_modeswitch and wvdial. This can be implemented in Python via shelling out to scripts or via direct wrappers to the tools if they exist. The main challenge is setting up the correct permissions on the webapp run user (we don't want to run the app as root for security reasons). The integration with the frontend for a "sync now" button then becomes trivial since it reduces to calling a method from one of the Flask views.

If you or someone you know would be interested in taking on this work, let me know.

**Comment by aidan-fitz** [comment]
I just created a [prototype] for the emergency mode button in Figma. Please try it and let me know if you have any feedback.

[prototype]: https://www.figma.com/proto/eMtetwKUE6bYpFk2fA9hD0ar/opwen-webapp-emergency-connect?node-id=0%3A1&scaling=scale-down

**Comment by c-w** [comment]
What about putting it into the nav where we currently have all the actions?

https://github.com/ascoderu/opwen-webapp/blob/c3544a8f1cdba3e2f44330ad30b754ee10038b0d/opwen_email_client/webapp/templates/macros/nav.html#L49-L74

E.g. an admin currently accesses the user blocking feature on the users page from the nav.

https://github.com/ascoderu/opwen-webapp/blob/c3544a8f1cdba3e2f44330ad30b754ee10038b0d/opwen_email_client/webapp/templates/users.html#L30-L43

Alternatively an admin-only button on the email page could also work from a conceptual grouping point of view.

https://github.com/ascoderu/opwen-webapp/blob/c3544a8f1cdba3e2f44330ad30b754ee10038b0d/opwen_email_client/webapp/templates/email.html#L6-L9

Or we add all the admin functionality to a page like home and move them to a dedicated admin page when there are enough features.

https://github.com/ascoderu/opwen-webapp/blob/c3544a8f1cdba3e2f44330ad30b754ee10038b0d/opwen_email_client/webapp/templates/home.html#L22-L36

**Comment by aidan-fitz** [comment]
I realized that it should prompt the device admin to set a time limit after which emergency mode automatically turns off. When people are distressed, they are likely to forget to turn it off manually, which could cost them a lot of money.

**Comment by aidan-fitz** [comment]
I was just prototyping the button by itself, but we can put it anywhere. I imagine it would sit in a settings page.

**Comment by c-w** [comment]
> I realized that it should prompt the device admin to set a time limit after which emergency mode automatically turns off. When people are distressed, they are likely to forget to turn it off manually, which could cost them a lot of money. (Update: just added this feature.)

We can solve this by having the "emergency mode" be a single sync invocation which sends/receives email starting the connection before sync and tearing it down after the sync is done. If another sync is desired, the admin can click the button again.

**Comment by aidan-fitz** [comment]
I'm imagining that emergency mode would be used in situations where people need to be receiving emergency broadcasts or sending messages to peers in real time, so my idea was that the device be continuously connected (or sync every few minutes) during the disaster event. Is it possible to use a WebSocket?

**Comment by c-w** [comment]
That would be a much larger change that doesn't fit well with what's currently in the project. A bit of Javascript that hits the button every N minutes/seconds would likely be a more pragmatic approach.

**Comment by nzola** [comment]
I plan to implement the Lokole in the Congo DRC, Africa in July 2019. I am sure the Admins who will be in charge of the Lokoles would love to see the emergency sync or continuous connection feature. 
As continuous connection future will create a major change on the software, I suggest to implement only the admin feature; sync and user management for blocking or unblocking etc.. This permission be given to the Admin only. And also the admin can be alternate within the networks if needed; only one person within the network ad a time.

**Comment by nzola** [comment]
One more think for the admin to do. To have a button to set the APN, PWD and USER NAME of the local mobile provider.

**Comment by c-w** [comment]
@nzola Created a new issue for the task above: https://github.com/ascoderu/opwen-webapp/issues/70

**Comment by c-w** [comment]
A settings page was implemented as part of https://github.com/ascoderu/opwen-webapp/pull/73 so the "sync now" button can be placed there.

**Comment by c-w** [comment]
@adamsclafani Here's some more information on the implementation:

1. We need a sync button in one of the templates (e.g. `settings.html`) so that the user can invoke the sync. This can be very straight forward: just a `<a>` tag which executes a get request to `/admin/sync`.

2. The `/admin/sync` route calls the `sync` function in `views.py` which in turn calls the `SyncEmails` action where the logic happens.

3. The current `SyncEmails` implementation has all the required logic to package up the pending emails and send them off to the remote server. It assumes that an internet connection already exists.

4. The internet connection is currently set up the `sync.sh` script that gets generated by `setup/setup-lokole.sh` ([here](https://github.com/ascoderu/opwen-webapp/blob/8799360726f17dbcb35bceece020c379d47b9ad8/setup/setup-lokole.sh#L588-L680)). There are a couple of steps to establish the internet connection:
  i. Find out the modem that's currently plugged in using `lsusb`.
  ii. Call `usb_modeswitch` to configure the modem if it's not already configured.
  iii. Call `wvdial` in the background to create the internet connection. Once the connection is no longer required, kill the pid of the wvdial process to shut down the connection.

Here's how I'd go about implementing the changes:

a. Make one new action: `StartInternetConnection`. Wrap the `SyncEmails` action with the new action, something like this:

```py
sync_emails = SyncEmails(...)
start_internet_connection = StartInternetConnection(...)

with start_internet_connection():
    sync_emails()
```

b. The example above shows how a [contextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) can be used to cleanly execute some code when the internet connection exits successfully or with an error, a bit like a try-finally block that the caller can't remember not to execute. This would be implemented somewhat like this:

```py
from contextlib import contextmanager

class StartInternetConnection(object):
    # ... other code here

    @contextmanager
    def __call__(self):
        # insert code to establish the connection here
        try:
            yield
        finally:
            # insert code to tear down the connection here
```

c. The code to establish the connection will be a bit more gnarly and will likely involve shelling-out to `usb_modeswitch` via [check_call](https://docs.python.org/3/library/subprocess.html#subprocess.check_call), running wvdial in the background via [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen). I wonder if the `lsusb` piece will have to be a shell-call via [check_output](https://docs.python.org/3/library/subprocess.html#subprocess.check_output) or whether [pyudev](https://pyudev.readthedocs.io/en/latest/) would be neater (I haven't used the latter so only look into it if you have an active interest in this area since the hacky subprocess call will likely be faster to implement as it's a straight port from the bash script).

d. The code to tear down wvdial should be fairly simple since the process object returned by Popen contains a [terminate](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.terminate) method.

**Comment by nzola** [comment]
Is this to say Sync Now is implemented?

- Nzola

**Comment by c-w** [comment]
Yes, it is. I still have to verify it end-to-end though.
